### PR TITLE
Remove redundant primitive types from intersections

### PR DESCRIPTION
**Pull request by ahejlsberg**
This PR removes primitive types `string`, `number`, and `symbol` from intersections that also contain literal types from the same domain. For example `string & 'a'` is reduced to just `'a'`. Likewise, `10 | number` is reduced to just `10`.

The PR eliminates the `TypeIncludes` enum from the type checker in favor of just using `TypeFlags`. This allows us to more efficiently compute the combined flags during the construction of a union or intersection type.

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -8270,13 +8243,27 @@ namespace ts {
...
         // types, duplicates are removed, and nested types of the given kind are flattened into the set.
-        function addTypesToIntersection(typeSet: Type[], includes: TypeIncludes, types: Type[]) {
+        function addTypesToIntersection(typeSet: Type[], includes: TypeFlags, types: Type[]) {
             for (const type of types) {
                 includes = addTypeToIntersection(typeSet, includes, getRegularTypeOfLiteralType(type));
             }
             return includes;
         }
 
+        function removeRedundantPrimtiveTypes(types: Type[], includes: TypeFlags) {
```
Comment (kpdonn) [inline src/compiler/checker.ts]: Typo: should be `removeRedundantPrimitiveTypes`

File: tests/cases/conformance/types/intersection/intersectionReduction.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
+// @strict
```
[the remark this answers is not in the document]
Comment (ahejlsberg) [reply tests/cases/conformance/types/intersection/intersectionReduction.ts]: I don't think so. We have hundreds of tests that include `// @strict: true` and they definitely behave differently if you remove the comment.

**Comment by mhegazy** [comment]
Fixes https://github.com/Microsoft/TypeScript/issues/9410
Fixes https://github.com/Microsoft/TypeScript/issues/23651

File: tests/cases/conformance/types/intersection/intersectionReduction.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (weswigham) [reply tests/cases/conformance/types/intersection/intersectionReduction.ts]: ....this comment isn't `// @strict: true`, it's just `// @strict`. This comment does nothing.

**Pull request merged**